### ISSUE TITLE
Surface layers should support `.properties`/`.features`

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari#5205]
## 🚀 Feature

Add  `.properties`/`.features` attributes to `Surface` layers.

## Motivation

Surface properties can be quantified.

## Pitch

We're working with some quantiative measurements of triangles in surface layers. In order to post-process them, it would be cool if `Surface`-layers supported `.properties`/`.features` just like `Labels` layers do.

CC @jo-mueller @kevinyamauchi @zoccoler @alisterburt @martinschatz-cz

See also
* https://github.com/haesleinhuepf/napari-process-points-and-surfaces/pull/15
* https://github.com/haesleinhuepf/napari-skimage-regionprops/pull/32
* https://github.com/haesleinhuepf/napari-accelerated-pixel-and-object-classification/pull/24

## Additional context

https://twitter.com/AlisterBurt/status/1579724195340357632?s=20&t=tpnb2D5NKFMnxry43d9cew

**Comment by jo-mueller** [comment]
Hi all,

I see that this issue hasn't received much traction, but I'd still be really interested in having it. As @haesleinhuepf points out, working with vertex-wise features is a quite common use-case. If someone could point me to the right direction I could also try myself at implementing it.

**Comment by alisterburt** [comment]
Hey @jo-mueller - sorry there wasn't much interest/activity here. I think there's a vicious cycle at play, the layer isn't great so not many people use it so no improvements so the layer isn't great...

I think this would be a great feature but if you need it you might be the only one with the drive to push for it in the near term. Maybe @andy-sweet could outline what it might take to get this done? I'm more than happy to test/review any PRs 🙂

**Comment by jni** [comment]
This is part of #2866 which is a priority on our [roadmap](https://github.com/orgs/napari/projects/24/views/2?pane=info) — though I feel like I didn't do a good enough job stressing its importance to the rest of the team. However @andy-sweet and I are pretty swamped and we could certainly use help from other community members...!

**Comment by jo-mueller** [comment]
Thanks for the prompt replies! For what it's worth, I'm at least glad that people would like to have it in the first place.

I may have *some* time in the summer to try myself at it but I'd definitely need a few directions. Looking forward to pointers!

**Comment by jni** [comment]
yeah I'm currently swamped with SciPy 2023 and tbh my summer looks crazy but I would indeed like to pair with you on this at some point. In the meantime @andy-sweet's [hackmd on the topic](https://github.com/napari/napari/issues/2866#issuecomment-1360790926) is essential reading. 😃

**Comment by jo-mueller** [comment]
Tbh, as a first step it would already enough for me if napari accepted something like

```
viewer.add_surface(data, features=some_dataframe)
```

without actually doing something with the passed dataframe besides attaching it to the layer. Being able to set the face_color with `viewer.add_surface(data, features=some_dataframe, face_color='column_name')` would be nice but I think at the moment not ultimately necessary, as some visualization plugins out there can maybe already handle this.

The fun part is that this already works (although it's probably not intended use):

```
viewer.layers["some_surface_layer"].features = some_dataframe
```

**Comment by andy-sweet** [comment]
Agreed in the long term, we should add `features` to `Surface` and make things consistent. I think if you're just looking to add that and don't need features-derived style/color, that's pretty doable. The big question for surfaces is if we want separate feature tables for vertices and triangles. At a minimum, I think we should have one for vertices, but the issue description also seems to want triangles.

> The fun part is that this already works (although it's probably not intended use):
> 
> ```
> viewer.layers["some_surface_layer"].features = some_dataframe
> ```

Another workaround is to use

```python
viewer.layers['some_surface_layer'].metadata['features'] = some_dataframe
```

That's more verbose, but one advantage is that the `metadata` attribute is expected by reader/writer plugins. Though the specific keys in the `metadata` dict are unstructured. So if you're just looking to store some values and maybe write them out, that might be a better option.

**Comment by jo-mueller** [comment]
@andy-sweet thanks for commenting on this. I think in the mid-future, having vertex-wise features would definitely enough for our (mine and @haesleinhuepf 's) usecases.

I guess a key change in supporting the features table (beyond just accepting it when passing it to the viewer) would be to deprecate the values entry in the surface tuple (`(vertices, faces, values)` which currently determines the rendering of colors on surfaces in napari...

**Comment by kevinyamauchi** [comment]
This would also be useful for me. Once https://github.com/napari/napari/pull/5477 (properties deprecation) goes in, I think we could add vertex features without having to also supporting the old properties API, which is nice.

**Comment by andy-sweet** [comment]
> This would also be useful for me. Once #5477 (properties deprecation) goes in, I think we could add vertex features without having to also supporting the old properties API, which is nice. I would propose to first just add a `vertex_features` property and then in a follow up PR hook up the coloring to the features table.
> 
> @andy-sweet, would this be disruptive to the broader properties -> features transition?

That sounds good to me. So does deprecating `values` in order to make `data` simpler (i.e. just vertices and faces).

**Comment by kevinyamauchi** [comment]
> So does deprecating values in order to make data simpler (i.e. just vertices and faces).

Would we replace it with something like `vertex_color` that behaves like `face_color` in the points layer? That is, one could directly specify the colors or map them against vertex features table?

**Comment by andy-sweet** [comment]
> > So does deprecating values in order to make data simpler (i.e. just vertices and faces).
> 
> Would we replace it with something like `vertex_color` that behaves like `face_color` in the points layer? That is, one could directly specify the colors or map them against vertex features table?

Yes. That's what I was thinking at a high level. But I'm not the most familiar with `Surface` and haven't thought through all the details/implications yet.

**Comment by zoccoler** [comment]
Hi guys,

A little bit side-topic, but still related, has anyone ever considered adding a `.properties`/`.features` to the [`Image layer`](https://napari.org/stable/api/napari.layers.Image.html) as well?

My use case is that I am working with raw FLIM images and I am considering storing the photon count dimension in a table-like structure.

**Comment by jo-mueller** [comment]
> Would we replace it with something like vertex_color that behaves like face_color in the points layer? That is, one could directly specify the colors or map them against vertex features table?

I'd say `vertex_color` does make sense to me also in the context of verbosity - it's clear that it's a vertex-wise coloring/feature table then. This would also leave some room to later introduce `face_color` or `edge_color` if such kinds of features should ever be supported in the future (but for which I don't see a super clear usecase, tbh).

@zoccoler would `features` for an Image then be pixel-wise features?

**Comment by zoccoler** [comment]
> @zoccoler would `features` for an Image then be pixel-wise features?

In my use-case yes.

**Comment by andy-sweet** [comment]
> Hi guys,
> 
> A little bit side-topic, but still related, has anyone ever considered adding a `.properties`/`.features` to the [`Image layer`](https://napari.org/stable/api/napari.layers.Image.html) as well?
> 
> My use case is that I am working with raw FLIM images and I am considering storing the photon count dimension in a table-like structure.

I think that has been discussed before in meetings (and maybe on github/zulip). After a quick search, I couldn't find an existing issue, so it could be helpful to create a separate explicit issue that to track the discussion and go into more details.

From my perspective, for points, shapes, vectors, surfaces, and tracks layers, `data` describes the sparse geometry/domain coordinates of the layer, whereas `features` contains a row for each element (e.g. a point) with associated values.

For an image layer, the geometry is dense with regularly spaced pixels that directly express values, so this use of `features` would be slightly different to the other layers, which is why a separate discussion could help. The labels layer also uses `features` in a special way (where each row is a label value), so some layer-specific definitions are fine.

**Comment by jo-mueller** [comment]
@andy-sweet this hasn't been revisited in a while so I was wondering whether this is currently being worked on?

If not, I could try my hand at implementing a fix that would allow features to be passed to the instantiation of a surface layer without throwing an error, which would really help me big time.

**Comment by andy-sweet** [comment]
@jo-mueller : I'm not currently working on it and I'm pretty busy this week. But I should have enough time to review stuff over the next few weeks and I'd be happy to take a look at anything you try. I have rough plans to try to finish off #5477 by the end of the year, but I need to pick some momentum back up there and take it over the line.

I think there's general support for adding this to the surface layer, but based on the conversation above there are a few tricky details.

If just adding `features` (or `vertex_features`) as a parameter/attribute and not worrying about the other issues discussed (deprecating values in data, hooking up to color maps) is enough for now, I think that should be fairly uncontentious.

**Comment by jo-mueller** [comment]
Thanks for the feedback! I'd say my list of priorities would be about this:

1. Surface layers not throwing an error anymore if features are passed
2. Coloring surfaces by a column of the features. The `values` entry of the surface layer tuple could, for the time being, simply be copied to the feature table before it is actually removed.
3. Removing the `values` entry in the surface layer.

**Comment by andy-sweet** [comment]
Sounds like a good plan. I'm particularly interested in helping out with (2), especially if [`ColorEncoding`](https://github.com/napari/napari/blob/6e0d23a595f2cd7ecf7ded62b0923173f720be40/napari/layers/utils/color_encoding.py#L32) can be used to handle the bulk of the implementation (though we don't necessarily have to expose it as part of the API like in [`TextManager`](https://github.com/napari/napari/blob/6e0d23a595f2cd7ecf7ded62b0923173f720be40/napari/layers/utils/text_manager.py#L84)).

**Comment by jo-mueller** [comment]
I was just about to follow up on this issue and realized that it has long been implemented [here](https://napari.org/stable/gallery/surface_texture_and_colors.html) I think this and https://github.com/napari/napari/issues/5261 can be closed 👍

**Comment by psobolewskiPhD** [comment]
@jo-mueller If i'm following, this is possibly closed by https://github.com/napari/napari/pull/6515 and https://github.com/napari/napari/pull/5642 ?
cc: @brisvag

**Comment by jo-mueller** [comment]
Yes 👍 Maybe @haesleinhuepf can confirm but I think that's what we had in mind, originally.

**Comment by brisvag** [comment]
Is this actually working? I'm not sure features are actually linked up?

**Comment by jo-mueller** [comment]
@brisvag I just gave it a look and a small piece may still be missing. I thought the new attribute `vertex_values` could be set to a feature of the features table, but it's just a view on the value element of the surface tuple (`surface_tuple[2])`, is it?

In any case, the missing bit would be in my eyes, that `vertex_colors` would accept a `str` as input as well and which should then color the surface accordingly. That's probably functionality that exists for other layers as well, so I could try myself at linking this up?

**Comment by brisvag** [comment]
Exactly! You would have to hook up the `_FeatureTable` with Manager classes such as `ColorManager` and `TextManager` (check out how `Points` implements it for an example).
Feel free to take a stab at it, and ask if you need help!

**Comment by jo-mueller** [comment]
@brisvag Ok, I'm at the point where I can instatiate a surface layer like this:

```python
viewer.add_surface(some_mesh, name='test', opacity=0.5, features=features, vertex_colors='feature2', colormap='plasma')
```

and the surface is correctly colored using the respective column in the features table. My next issue would be how to link up the layer controls (e.g., colormap, contrast limits & gamma) so that they affect the layer display. I *guess* that I need to hook up the `ColorManager` to the respective events but I'm not sure whether that's happening in the respective layer file (i.e., `surface.py`).

**Comment by brisvag** [comment]
Awesome, you're already basically all the way there! If you open a PR from your branch we can continue the discussion there with the context of the diff, which makes it a lot easier to give feedback :)
